src/cryptonote_core/quorum_cop.cpp

```diff
@@ -0,0 +1,221 @@
+        continue;
+      }
+
+      size_t my_index_in_quorum = 0;
+      if (!participates_in_quorum(state->quorum_nodes, my_pubkey, &my_index_in_quorum))
+      {
+        continue;
+      }
+
+      for (size_t node_index = 0; node_index < state->nodes_to_test.size(); ++node_index)
+      {
+        const crypto::public_key &node_key = state->nodes_to_test[node_index];
+        bool vote_off_node = false;
+
+        CRITICAL_REGION_LOCAL(m_lock);
+        auto const it = m_uptime_proof_seen.find(node_key);
+        if (it == m_uptime_proof_seen.end())
+        {
+          vote_off_node = true;
+        }
+        else
+        {
+          size_t time_since_proof = now - it->second;
+          vote_off_node           = (time_since_proof > UPTIME_PROOF_MAX_TIME_IN_SECONDS) ? true : false;
+        }
+
+        if (!vote_off_node)
+          continue;
+
+        loki::service_node_deregister::vote vote = {};
+        vote.block_height        = m_last_height;
+        vote.service_node_index  = node_index;
+        vote.voters_quorum_index = my_index_in_quorum;
+        vote.signature           = loki::service_node_deregister::sign_vote(vote.block_height, vote.service_node_index, my_pubkey, my_seckey);
+
+        cryptonote::vote_verification_context vvc = {};
+        if (!m_core.add_deregister_vote(vote, vvc))
+        {
+          if (vvc.m_invalid_block_height)
+            LOG_ERROR("block height was invalid: " << vote.block_height);
+
+          if (vvc.m_voters_quorum_index_out_of_bounds)
+            LOG_ERROR("voters quorum index specified out of bounds: " << vote.voters_quorum_index);
+
+          if (vvc.m_service_node_index_out_of_bounds)
+            LOG_ERROR("service node index specified out of bounds: " << vote.service_node_index);
+
+          if (vvc.m_signature_not_valid)
+            LOG_ERROR("signature was not valid, was the signature signed properly?");
+        }
+      }
+    }
+
+    prune_uptime_proof_below(execute_justice_from_height);
+  }
+
+  static crypto::hash make_hash(crypto::public_key const &pubkey, uint64_t timestamp)
+  {
+    // TODO(doyle): Errrr clarify what this is. Looks like crypto::cn_fast_hash should be sizeof(buf)
+    char buf[44] = "SUP"; // Meaningless magic bytes
+    crypto::hash result;
+    memcpy(buf + 4, reinterpret_cast<const void *>(&pubkey), sizeof(pubkey));
+    memcpy(buf + 4 + sizeof(pubkey), reinterpret_cast<const void *>(&timestamp), sizeof(timestamp));
+    crypto::cn_fast_hash(buf, sizeof(buf), result);
+
+    return result;
+  }
+
+  bool quorum_cop::handle_uptime_proof(uint64_t timestamp, const crypto::public_key& pubkey, const crypto::signature& sig)
+  {
+    uint64_t now = time(nullptr);
+
+    if ((timestamp < now - UPTIME_PROOF_BUFFER_IN_SECONDS) || (timestamp > now + UPTIME_PROOF_BUFFER_IN_SECONDS))
+      return false;
+
+    if (!m_service_node_list.is_service_node(pubkey))
+      return false;
```

Comment:
The only dependency on service_node_lists. Once we get db storage of service node lists we can remove this and check if the pubkey belongs to some range of heights.

I can imagine we'll want to store uptime proof over some range of heights so we have the tools to provide statistics on the network.

Reply:
Noted